### PR TITLE
Modulegraph: print errors to logger.

### PR DESCRIPTION
**Pull request by xoviat**
Once again, this will be rebased on develop after other PRs are merged.

### CODE REVIEW

File: PyInstaller/lib/modulegraph/modulegraph.py [hunk 1]
```diff
@@ -9,10 +9,7 @@
 """
 from __future__ import absolute_import, print_function
 
-import pkg_resources
-
 import ast
-import dis
```
Comment (htgoebel) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Unrelated change

File: PyInstaller/lib/modulegraph/modulegraph.py [hunk 2]
```diff
@@ -1105,7 +1103,6 @@ def createNode(self, cls, name, *args, **kw):
 
         return m
 
-
```
Comment (htgoebel) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Don't remove these empty lines, this is PEP 8 style.

File: PyInstaller/lib/modulegraph/modulegraph.py [hunk 3]
```diff
@@ -1125,6 +1122,19 @@ def __init__(self, path=None, excludes=(), replace_paths=(), implies=(), graph=N
...
+        Print a debug message with the given level
+        """
+        if s:
+            message = "%s %s" % (s, ' '.join(map(repr, args)))
+            if level <= 1:
+                logger.warning(message)
+            elif level <= 2:
+                logger.info(message)
+            elif level <= 3:
+                logger.debug(message)
```
Comment (htgoebel) [inline PyInstaller/lib/modulegraph/modulegraph.py]: What happens, if level is e.g. `4`, as seen quite often in this very module?

File: PyInstaller/lib/modulegraph/modulegraph.py [hunk 4]
```diff
@@ -2108,8 +2116,7 @@ def _safe_import_module(
         self.msgout(3, "safe_import_module ->", module)
         return module
 
-
-    #FIXME: For clarity, rename method parameters to:
+    # FIXME: For clarity, rename method parameters to:
```
Comment (htgoebel) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Unrelated change (and unnecessary)

**Comment by xoviat** [comment]
Honestly I hadn't planned for this to be merged yet because this commit appears to create a merge conflict if the order of the commits is changed.

File: PyInstaller/lib/modulegraph/modulegraph.py (hunk 3 above)
Comment (xoviat) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Logging prints out many messages with the level set to 3. I made a judgement call that raising the level to 4 would print out too many messages. My position here is indefensible, so feel free to change the numbers if you want.

File: PyInstaller/lib/modulegraph/modulegraph.py (hunk 1 above)
Comment (xoviat) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Pycharm cleans up unused imports.

**Comment by htgoebel** [comment]
> I made a judgement call that raising the level to 4 would print out too many messages.

That's fair (but you should have written it into a comment :-)

But then: how will we ever get theses "level 4" messages? I found them useful when debugging the modulegraph test-suite at travis.

I was thinking about moving this change to `depend.analysis.PyiModuleGraph` anyway. If we put it here, debugging the deep internals of modulegraph would require configuring the logger here. If we move it to PyiModuleGraph we would get the same effect when running `pyinstaller` and the pyinstaller test-suite, but will still be able to easily enable debugging for the modulegraph test-suite (and also keep the indention-levels)

What do you think? If you agree would you please implement this change (mind adding the comment ;-) and update the pull-request to only include this one commit? I think we can use the existing logger in `depend.analysis`. Thanks.

> create a merge conflict if the order of the commits is changed

I want to close the easy ones first re-establish an overview over the pull-requests. Maybe this is just since it includes unrelated changes?

File: PyInstaller/lib/modulegraph/modulegraph.py (hunk 1 above)
Comment (htgoebel) [inline PyInstaller/lib/modulegraph/modulegraph.py]: Are you used to `gui gui` and `git citool` already? These allow to select even single lines for inclusion or exclusion from the patch. I really like this mighty tools.

**Comment by xoviat** [comment]
> What do you think? If you agree would you please implement this change (mind adding the comment ;-) and update the pull-request to only include this one commit? I think we can use the existing logger in depend.analysis. Thanks.

I think that's a good idea. I'll update the pull request with the new commit.

File: PyInstaller/depend/analysis.py [hunk 5]
```diff
@@ -100,6 +101,32 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+            elif lvl <= 4:
+                level = DEBUG
+
+        try:
+            fn, lno, func, sinfo = logger.findCaller()
+        except ValueError:  # pragma: no cover
+            fn, lno, func = "(unknown file)", 0, "(unknown function)"
+
+        record = logger.makeRecord(
+            logger.name, level, fn, lno, msg, args, None, func, None, sinfo)
```
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: `makeRecord is different on py27`

Comment (xoviat) [inline PyInstaller/depend/analysis.py]: `Logger.makeRecord(name, lvl, fn, lno, msg, args, exc_info, func=None, extra=None)`

**Comment by xoviat** [comment]
So it looks like py.test cannot handle too many messages.

**Comment by xoviat** [comment]
It does appear that the messages are printing out quite nicely. Would you like to set the verbosity level back to 3 (which apparently doesn't crash py.test), or do something else?

File: PyInstaller/depend/analysis.py [hunk 6]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+        level = None
+
+        # Note: these levels are completely arbitrary and may be adjusted if needed.
+        if s:
+            msg = "%s %s" % (s, ' '.join(map(repr, args)))
+            if lvl <= 1:
+                level = WARNING
+            elif lvl <= 2:
+                level = INFO
+            elif lvl <= 4:
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: What about using a mapping here:
```
LEVEL_MAPPING = {0: WARNING, 1:WARNING, 2: INFO, 3: DEBUG}
…
level = LEVEL.MAPPING.get(level)
```
This will make any other value (esp `4`) into `None` and negative values should not occur anyway. Or do they?

You could also save globally importing WARNING, INFO and DEBUG.(see line 38).

File: PyInstaller/depend/analysis.py [hunk 7]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+                level = DEBUG
+
+        if level is None:
+            return
+
+        try:
+            fn, lno, func, sinfo = logger.findCaller()
+        except ValueError:  # pragma: no cover
+            fn, lno, func = "(unknown file)", 0, "(unknown function)"
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: Just set `sinfo = None` here, no need to pre-init the variable above.

File: PyInstaller/depend/analysis.py [hunk 8]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+
+        if is_py2:
+            record = logger.makeRecord(
+                logger.name, level, fn, lno, msg, [], None, func, None)
+        else:
+            record = logger.makeRecord(
+                logger.name, level, fn, lno, msg, [], None, func, None, sinfo)
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: Given the mass of messages modulegraph produces and the mass of iterations on modulegraph we do, we should not test this again and again for every message.

I suggest distinguishing this once at class level. Either by defining `PyiModuleGraph.__makeLogRecord` or even by duplicating the whole method:
```
if is_py2:
   def _msg(): …
else:
  def _msg(): …
```

File: PyInstaller/depend/analysis.py [hunk 9]
```diff
@@ -35,18 +35,19 @@
...
-from ..lib.modulegraph.modulegraph import ModuleGraph
-from ..lib.modulegraph.find_modules import get_implies
 from ..compat import importlib_load_source, is_py2, PY3_BASE_MODULES,\
         PURE_PYTHON_MODULE_TYPES, BINARY_MODULE_TYPES, VALID_MODULE_TYPES, \
         BAD_MODULE_TYPES, MODULE_TYPES_TO_TOC_DICT
-from .. import HOMEPATH, configure
+from ..lib.modulegraph.find_modules import get_implies
+from ..lib.modulegraph.modulegraph import ModuleGraph
 from ..utils.hooks import collect_submodules, is_package
+from ..utils.misc import load_py_data_struct
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: This is a non-functional change, please put it into a separate commit.

File: PyInstaller/depend/analysis.py [hunk 10]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
             os.path.join(self._homepath, 'PyInstaller', 'loader', 'rthooks.dat')
         )
 
+    def msg(self, *args, **kwargs):
+        self._msg(*args, **kwargs)
+
+    def _msg(self, lvl, s, *args):
```
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: The stack depth needs to be set correctly in order to determine where the call to `msg` came from. This could be determined manually but logging has a facility built in if you set the correct stack depth.

File: PyInstaller/depend/analysis.py [hunk 11]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+                level = INFO
+            elif lvl <= 4:
+                level = DEBUG
+
+        if level is None:
+            return
+
+        try:
+            fn, lno, func, sinfo = logger.findCaller()
+        except ValueError:  # pragma: no cover
```
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: The is a rewrite of python's `_log` method, which handles this exception, so I assume it can.

File: PyInstaller/depend/analysis.py [hunk 12]
```diff
@@ -81,6 +82,10 @@ class PyiModuleGraph(ModuleGraph):
 
 
     def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
+        # Set logging methods so that the stack is correctly detected.
+        self.msgin = self.msg
+        self.msgout = self.msg
```
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: The comment is this:
```
        # Set logging methods so that the stack is correctly detected.
```

File: PyInstaller/depend/analysis.py (hunk 8 above)
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: According to my calculations (which admittedly may be incorrect), a function call takes 0.11 microseconds whereas an if statement takes 0.057 microseconds, so the if statement should be faster.

File: PyInstaller/depend/analysis.py (hunk 9 above)
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: Done.

File: PyInstaller/depend/analysis.py [hunk 13]
```diff
@@ -100,6 +105,43 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+            fn, lno, func = "(unknown file)", 0, "(unknown function)"
+
+        if is_py2:
+            record = logger.makeRecord(
+                logger.name, level, fn, lno, msg, [], None, func, None)
+        else:
+            record = logger.makeRecord(
+                logger.name, level, fn, lno, msg, [], None, func, None, sinfo)
+
+        logger.handle(record)
```
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: Updated the docstring.

File: PyInstaller/depend/analysis.py (hunk 8 above)
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: :-) Well done! I should have concluded this be thinking, too, since the test only requires a variable look-up, while the call requires setting up a stack-frame.

File: PyInstaller/depend/analysis.py (hunk 12 above)
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: IC.

File: PyInstaller/depend/analysis.py (hunk 10 above)
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: I still don't understand, why we need this additional function `_msg`. Why can't this be done in `msg()` already?

File: PyInstaller/depend/analysis.py [hunk 14]
```diff
@@ -100,6 +103,45 @@ def __init__(self, pyi_homepath, user_hook_dirs=None, *args, **kwargs):
...
+        2. Generate the message format (the same format as ModuleGraph)
+        3. Find the caller, which is two functions above:
+            [3] caller -> [2] msg -> [1] _msg (here) -> [0] logger.findCaller
+        4. Create a logRecord with the caller's information.
+        5. Handle the logRecord.
+        """
+        try:
+            level = self.LOG_LEVEL_MAPPING[level]
+        except KeyError:
+            return
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: :-) Well done, this is more elegant than what I scratched.

**Comment by htgoebel** [comment]
> 	Modulegraph: logging: reduce verbosity for py2

What is the reasons for this?

File: PyInstaller/depend/analysis.py [hunk 15]
```diff
@@ -128,15 +128,20 @@ def _msg(self, level, s, *args):
...
-        try:
-            fn, lno, func, sinfo = logger.findCaller()
-        except ValueError:  # pragma: no cover
-            fn, lno, func, sinfo = "(unknown file)", 0, "(unknown function)", None
-
         if is_py2:
+            try:
+                fn, lno, func = logger.findCaller()
+            except ValueError:  # pragma: no cover
+                fn, lno, func = "(unknown file)", 0, "(unknown function)", None
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: This will fail: wrong number of values

File: PyInstaller/depend/analysis.py [hunk 16]
```diff
@@ -128,15 +128,20 @@ def _msg(self, level, s, *args):
...
             record = logger.makeRecord(
                 logger.name, level, fn, lno, msg, [], None, func, None)
         else:
+            try:
+                fn, lno, func, sinfo = logger.findCaller()
+            except ValueError:  # pragma: no cover
+                fn, lno, func, sinfo = "(unknown file)", 0, "(unknown function)", None
+            
             record = logger.makeRecord(
                 logger.name, level, fn, lno, msg, [], None, func, None, sinfo)
```
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: This will fail: sinfo no defined.

Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: Will not fail, this is Py 3 only.

File: PyInstaller/depend/analysis.py (hunk 15 above)
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: BTW, you're commenting on outdated changes.

**Comment by htgoebel** [comment]
Buh, this produces **lot** of messages:

levels | lines
-------|------
no logging | 561
0 and 1 | 579
up to 2 | 1700
up to 3 | 13266
up to 4 | 32486

And I'm afraid, the output above "1" is not of much use, except for digging *deep* into modulegraph: Level 2 already brings this level of details:
```
230 INFO: load_module 'codecs' 'fp' '/tmp/v27/lib/python2.7/codecs.py'
252 INFO: load_module '__builtin__' None None
252 INFO: load_module -> BuiltinModule('__builtin__',)
252 INFO: load_module 'sys' None None
252 INFO: load_module -> BuiltinModule('sys',)
253 INFO: load_module '_codecs' None None
253 INFO: load_module -> BuiltinModule('_codecs',)
```

So I propose to only use messages up to level "1", even if Pyinstaller is logging to at DEBUG level.

To get debug output for modulegraph, I see these options:
1. Don't provide support (above level 1). If this is required, create test-case for underlying modulegraph and debug this.
2. Use the [already implemented way of setting MODULEGRAPH_DEBUG](https://github.com/pyinstaller/pyinstaller/blob/v3.2.1/PyInstaller/depend/analysis.py#L524) to either configure the logger or to change LOG_LEVEL_MAPPING.
3. Find some other way to configure the logger for PyiModuleGraph.

What do you think?

File: PyInstaller/depend/analysis.py (hunk 15 above)
Comment (htgoebel) [inline PyInstaller/depend/analysis.py]: If this commit is outdated, why is it shown in https://github.com/pyinstaller/pyinstaller/pull/2433/commits then?

**Comment by xoviat** [comment]
We can provide a `--modulegraph-debug` option on the command line that changes the mapping.

**Comment by htgoebel** [comment]
> We can provide a --modulegraph-debug option on the command line that changes the mapping.

How should it work? Passing the maximum moduegraph level to be mapped (and thus logged)?

I'd rather not clutter the command line options with this. It is very, vary rarely used except by developer. IMHO we can do the same by using MODULEGRAPH_DEBUG – of course we should then document it :)

**Comment by xoviat** [comment]
Scratch that. What I'm going to do is provide a new logging level called `TRACE` and log the messages on that level.

**Comment by xoviat** [comment]
Appears to work quite nicely!

File: PyInstaller/depend/analysis.py (hunk 15 above)
Comment (xoviat) [inline PyInstaller/depend/analysis.py]: The commits tab shows all commits, but some commits overwrite changes by previous commits.

**Comment by htgoebel** [comment]
Thanks for this update, this now look really good and in nearly ready to be merged. :-) Introducing TRACE is a good idea. I was about suggesting to move the TRACE to `log`, and you already did it. Great!

Only two smal qestions to answer:
1. Why is the an additional function `_msg()`, which `msg()` is calling without any changes. Can't we remove this extra call-level?
2. Regarging [catchlog] in setup.py: Please provide a comment stating what this is for, like "make py.test-catchlog only catch messages up to DEBUG level"

**Comment by xoviat** [comment]
> Why is the an additional function _msg(), which msg() is calling without any changes. Can't we remove this extra call-level?

Look at the docstring:


>        3. Find the caller, which is two functions above:
>             [3] caller -> [2] msg -> [1] _msg (here) -> [0] logger.findCaller

**Comment by htgoebel** [comment]
> Find the caller, which is two functions above:

But if we remove the additional `_msg()` call-level, it's only one level up. Thus this comment's meaning is not obvious for me.
Why should this matter? Does findCaller expect it to be *exactly* two levels above?

**Comment by xoviat** [comment]
> Does findCaller expect it to be exactly two levels above?

Correct. Options are we can use the logging `findCaller` or we can write our own function that is dependent on implementation-specific functionality for a small performance gain.

**Comment by xoviat** [comment]
Actually, scratch that. We can wrap findCaller in a function, which would yield the best of both worlds.

**Comment by xoviat** [comment]
Done and Done.

**Comment by htgoebel** [comment]
> Modulegraph: logging: rework stack order.

Yeah, this is a nice one: Pushing the addition all after checking whether the additional level is required at all.

Thanks.

**Comment by htgoebel** [comment]
Thanks for this pull-request.
